### PR TITLE
Plugin implementation

### PR DESCRIPTION
**Pull request by alalamav**
- All of the plugin's functionality is included in these changes.
- The main commits contain modifications to tun2socks [(commit)](https://github.com/uProxy/cordova-plugin-tun2socks/commit/39929e5f4751662286a0020a9d7f30975b542fa6) and Psiphon [(commit)](https://github.com/uProxy/cordova-plugin-tun2socks/commit/f1b880f002ea162cfcfcd60f94c1688a508c624a).
- Once the code is reviewed, the plugin's import path should be updated  in uProxy's Gruntfile.

---

This change is [Reviewable](https://reviewable.io/reviews/uproxy/cordova-plugin-tun2socks/3)

### CODE REVIEW

**Comment by bemasc** [comment]
Review status: 0 of 19 files reviewed at latest revision, 13 unresolved discussions.

---

_[plugin.xml, line 27 [r2]](https://reviewable.io:443/reviews/uproxy/cordova-plugin-tun2socks/3#-KNsAwGVx-Ae7G_FQQGE:-KNsAwGVx-Ae7G_FQQGF:-1294256198) ([raw file](https://github.com/uproxy/cordova-plugin-tun2socks/blob/55fa490af8903985f4212a350693068441273c22/plugin.xml#L27)):_

> ``` XML
> 
>     <config-file target="AndroidManifest.xml" parent="/manifest/application">
> ```

Nit: blank line

---

_[src/android/libs/jsocks.jar, line 0 [r2]](https://reviewable.io:443/reviews/uproxy/cordova-plugin-tun2socks/3#-KNsBeYdlgRqDSPxQYKm:-KNsBeYefgySJqg1O4-T:-952779712) ([raw file](https://github.com/uproxy/cordova-plugin-tun2socks/blob/55fa490af8903985f4212a350693068441273c22/src/android/libs/jsocks.jar#L0)):_
Obviously it's a bit unfortunate to be committing a JAR file.  How big is this file?  Do we have an alternative way to include it in the build?

---

_[src/android/org/uproxy/tun2socks/DnsResolverService.java, line 28 [r2]](https://reviewable.io:443/reviews/uproxy/cordova-plugin-tun2socks/3#-KNsByYEyme2yx6Vuezm:-KNsByYLyx-l9-I5alFu:-1657308043) ([raw file](https://github.com/uproxy/cordova-plugin-tun2socks/blob/55fa490af8903985f4212a350693068441273c22/src/android/org/uproxy/tun2socks/DnsResolverService.java#L28)):_

> ``` Java
> 
> // This class listens on a UDP socket for incoming DNS traffic, which is proxy'd
> // through SOCKS over TCP to Honest DNS.
> ```

Nobody calls it Honest DNS :)

---

_[src/android/org/uproxy/tun2socks/DnsResolverService.java, line 141 [r2]](https://reviewable.io:443/reviews/uproxy/cordova-plugin-tun2socks/3#-KNsDDztCarHvfdckaBq:-KNsDDzuJB8j6AvX-uiA:-1644806875) ([raw file](https://github.com/uproxy/cordova-plugin-tun2socks/blob/55fa490af8903985f4212a350693068441273c22/src/android/org/uproxy/tun2socks/DnsResolverService.java#L141)):_

> ``` Java
>           Log.i(LOG_TAG, "listening on " + udpSocket.getLocalSocketAddress().toString());
>           try {
>             udpSocket.receive(udpPacket);
> ```

What happens if someone sends a packet that's bigger than 512 bytes?

---

_[src/android/org/uproxy/tun2socks/DnsResolverService.java, line 199 [r2]](https://reviewable.io:443/reviews/uproxy/cordova-plugin-tun2socks/3#-KNsDhEb7lXCJuAvYXDg:-KNsDhEb7lXCJuAvYXDh:2079304989) ([raw file](https://github.com/uproxy/cordova-plugin-tun2socks/blob/55fa490af8903985f4212a350693068441273c22/src/android/org/uproxy/tun2socks/DnsResolverService.java#L199)):_

> ``` Java
>       } catch (Throwable e) {
>         if (!isInterrupted()) {
>           e.printStackTrace();
> ```

What state is the app now in?  Is it silently broken?

---

_[src/android/org/uproxy/tun2socks/DnsResolverService.java, line 208 [r2]](https://reviewable.io:443/reviews/uproxy/cordova-plugin-tun2socks/3#-KNsDy326NWVtB8voaD1:-KNsDy326NWVtB8voaD2:-748018645) ([raw file](https://github.com/uproxy/cordova-plugin-tun2socks/blob/55fa490af8903985f4212a350693068441273c22/src/android/org/uproxy/tun2socks/DnsResolverService.java#L208)):_

> ``` Java
>     }
> 
>     private boolean validateDnsRequest(DatagramPacket packet) {
> ```

Nit: Prefer `isValid` over `validate`.  Also, could this method be static?

---

_[src/android/org/uproxy/tun2socks/Tun2Socks.java, line 97 [r2]](https://reviewable.io:443/reviews/uproxy/cordova-plugin-tun2socks/3#-KNsHOMP4MqsrZFx3Vf6:-KNsHOMP4MqsrZFx3Vf7:898230277) ([raw file](https://github.com/uproxy/cordova-plugin-tun2socks/blob/55fa490af8903985f4212a350693068441273c22/src/android/org/uproxy/tun2socks/Tun2Socks.java#L97)):_

> ``` Java
>   }
> 
>   private boolean hasVpnService() {
> ```

Nit: I think it would be worth exposing this as a command.  Then maybe we wouldn't need to check the OS version in the JS code; we could just ask the plugin whether it's going to work.

---

_[src/android/org/uproxy/tun2socks/Tun2Socks.java, line 104 [r2]](https://reviewable.io:443/reviews/uproxy/cordova-plugin-tun2socks/3#-KNsHktXmO_LnnVnUFEV:-KNsHktXmO_LnnVnUFEW:555843179) ([raw file](https://github.com/uproxy/cordova-plugin-tun2socks/blob/55fa490af8903985f4212a350693068441273c22/src/android/org/uproxy/tun2socks/Tun2Socks.java#L104)):_

> ``` Java
>   protected boolean prepareVpnService() throws ActivityNotFoundException {
>     // VpnService: need to display OS user warning. If whole device
>     // option is selected and we expect to use VpnService, so the prompt here in
> ```

Nit: "show the prompt"?

---

_[src/android/org/uproxy/tun2socks/Tunnel.java, line 127 [r2]](https://reviewable.io:443/reviews/uproxy/cordova-plugin-tun2socks/3#-KNsIMSGz6-4BmLAUjqh:-KNsIMSHyqH4CK12EwB4:1094593170) ([raw file](https://github.com/uproxy/cordova-plugin-tun2socks/blob/55fa490af8903985f4212a350693068441273c22/src/android/org/uproxy/tun2socks/Tunnel.java#L127)):_

> ``` Java
>   // synchronized functions as stop() is synchronized and a deadlock is possible as callbacks
>   // can be called while stop holds the lock.
>   @TargetApi(Build.VERSION_CODES.LOLLIPOP)
> ```

Why lollipop?  Deserves a comment.

---

_[src/android/org/uproxy/tun2socks/TunnelManager.java, line 59 [r2]](https://reviewable.io:443/reviews/uproxy/cordova-plugin-tun2socks/3#-KNsIkwxJn4AZRKaY-Zb:-KNsIkwxJn4AZRKaY-Zc:832000065) ([raw file](https://github.com/uproxy/cordova-plugin-tun2socks/blob/55fa490af8903985f4212a350693068441273c22/src/android/org/uproxy/tun2socks/TunnelManager.java#L59)):_

> ``` Java
> 
>   // Implementation of android.app.Service.onStartCommand
>   @TargetApi(Build.VERSION_CODES.M)
> ```

Why M?  Needs comment.

---

_[src/badvpn/tun2socks/tun2socks.c, line 323 [r2]](https://reviewable.io:443/reviews/uproxy/cordova-plugin-tun2socks/3#-KNsJxSdQX-cotYwEzk6:-KNsJxSdQX-cotYwEzk7:-1838768796) ([raw file](https://github.com/uproxy/cordova-plugin-tun2socks/blob/55fa490af8903985f4212a350693068441273c22/src/badvpn/tun2socks/tun2socks.c#L323)):_

> ``` C
> 
> // Initializes a UDP socket, binds it locally and connects it to a local
> // address.
> ```

What's the return value?

---

_[src/badvpn/tun2socks/tun2socks.c, line 337 [r2]](https://reviewable.io:443/reviews/uproxy/cordova-plugin-tun2socks/3#-KNsJYfP5dao7dxYbGC9:-KNsJYfP5dao7dxYbGCA:343544511) ([raw file](https://github.com/uproxy/cordova-plugin-tun2socks/blob/55fa490af8903985f4212a350693068441273c22/src/badvpn/tun2socks/tun2socks.c#L337)):_

> ``` C
>         return 0;
>     }
>     // Bind to any local address in order to receive data
> ```

INADDR_ANY is not the same as "any local address".  It's more like "the ANY address".

---

_[src/badvpn/tun2socks/tun2socks.h, line 48 [r2]](https://reviewable.io:443/reviews/uproxy/cordova-plugin-tun2socks/3#-KNsKXRLQDihiFcBfeGB:-KNsKXRLQDihiFcBfeGC:-1715047403) ([raw file](https://github.com/uproxy/cordova-plugin-tun2socks/blob/55fa490af8903985f4212a350693068441273c22/src/badvpn/tun2socks/tun2socks.h#L48)):_

> ``` objc
> //#define OVERRIDE_DEST_ADDR "10.111.0.2:2000"
> 
> // maximum number of bytes for udp send/recv
> ```

Is this ever applied to UDP sending?

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/uproxy/cordova-plugin-tun2socks/3)_

**Comment by alalamav** [comment]
Review status: 0 of 19 files reviewed at latest revision, 13 unresolved discussions.

---

_[plugin.xml, line 27 [r2]](https://reviewable.io:443/reviews/uproxy/cordova-plugin-tun2socks/3#-KNsAwGVx-Ae7G_FQQGE:-KNsQezUcyBDiejM-FNC:-499146009) ([raw file](https://github.com/uproxy/cordova-plugin-tun2socks/blob/55fa490af8903985f4212a350693068441273c22/plugin.xml#L27)):_
<details><summary>

<i>Previously, bemasc (Benjamin M. Schwartz) wrote…</i></summary>



> Nit: blank line
> </details>
> Done.

---

_[src/android/libs/jsocks.jar, line  [r2]](https://reviewable.io:443/reviews/uproxy/cordova-plugin-tun2socks/3#-KNsBeYdlgRqDSPxQYKm:-KNsPgiPoR74izOorwec:-693046056) ([raw file](https://github.com/uproxy/cordova-plugin-tun2socks/blob/55fa490af8903985f4212a350693068441273c22/src/android/libs/jsocks.jar#L)):_
<details><summary>

<i>Previously, bemasc (Benjamin M. Schwartz) wrote…</i></summary>



> Obviously it's a bit unfortunate to be committing a JAR file.  How big is this file?  Do we have an alternative way to include it in the build?
> </details>
> The JAR is 395 KB. The source code is in SourceForge, but seems like it's not actively maintained. We could create our fork, and include the JAR ad hoc. But I don't know if this is a better alternative.

---

_[src/android/org/uproxy/tun2socks/DnsResolverService.java, line 28 [r2]](https://reviewable.io:443/reviews/uproxy/cordova-plugin-tun2socks/3#-KNsByYEyme2yx6Vuezm:-KNsQgFf4_3Os4CtYVhl:160412708) ([raw file](https://github.com/uproxy/cordova-plugin-tun2socks/blob/55fa490af8903985f4212a350693068441273c22/src/android/org/uproxy/tun2socks/DnsResolverService.java#L28)):_
<details><summary>

<i>Previously, bemasc (Benjamin M. Schwartz) wrote…</i></summary>



> Nobody calls it Honest DNS :)
> </details>
> lol

---

_[src/android/org/uproxy/tun2socks/DnsResolverService.java, line 141 [r2]](https://reviewable.io:443/reviews/uproxy/cordova-plugin-tun2socks/3#-KNsDDztCarHvfdckaBq:-KNsR-60EiHrEcKgAFwY:-2039840669) ([raw file](https://github.com/uproxy/cordova-plugin-tun2socks/blob/55fa490af8903985f4212a350693068441273c22/src/android/org/uproxy/tun2socks/DnsResolverService.java#L141)):_
<details><summary>

<i>Previously, bemasc (Benjamin M. Schwartz) wrote…</i></summary>



> What happens if someone sends a packet that's bigger than 512 bytes?
> </details>
> As discussed offline, the DNS over UDP limit is 512 bytes and it's very unlikely that a query will exceed this size. However, the response could have a size of 32767 (max UDP packet size). 
> I'll increase the buffer size to account for this.

---

_[src/android/org/uproxy/tun2socks/DnsResolverService.java, line 199 [r2]](https://reviewable.io:443/reviews/uproxy/cordova-plugin-tun2socks/3#-KNsDhEb7lXCJuAvYXDg:-KNsnTEYvOWJGCMXoPmh:677706293) ([raw file](https://github.com/uproxy/cordova-plugin-tun2socks/blob/55fa490af8903985f4212a350693068441273c22/src/android/org/uproxy/tun2socks/DnsResolverService.java#L199)):_
<details><summary>

<i>Previously, bemasc (Benjamin M. Schwartz) wrote…</i></summary>



> What state is the app now in?  Is it silently broken?
> </details>
> I've refactored this class to catch all exceptions, so this catch is no longer needed and there's no ambiguity about what went wrong.

---

_[src/android/org/uproxy/tun2socks/DnsResolverService.java, line 208 [r2]](https://reviewable.io:443/reviews/uproxy/cordova-plugin-tun2socks/3#-KNsDy326NWVtB8voaD1:-KNsRW8U2vKTWZllKEmO:-1335584080) ([raw file](https://github.com/uproxy/cordova-plugin-tun2socks/blob/55fa490af8903985f4212a350693068441273c22/src/android/org/uproxy/tun2socks/DnsResolverService.java#L208)):_
<details><summary>

<i>Previously, bemasc (Benjamin M. Schwartz) wrote…</i></summary>



> Nit: Prefer `isValid` over `validate`.  Also, could this method be static?
> </details>
> Done. It cannot because non-static nested classes are not allowed to declare static methods.

---

_[src/android/org/uproxy/tun2socks/Tun2Socks.java, line 104 [r2]](https://reviewable.io:443/reviews/uproxy/cordova-plugin-tun2socks/3#-KNsHktXmO_LnnVnUFEV:-KNsaST4PTdFtNnRI3gd:-499146009) ([raw file](https://github.com/uproxy/cordova-plugin-tun2socks/blob/55fa490af8903985f4212a350693068441273c22/src/android/org/uproxy/tun2socks/Tun2Socks.java#L104)):_
<details><summary>

<i>Previously, bemasc (Benjamin M. Schwartz) wrote…</i></summary>



> Nit: "show the prompt"?
> </details>
> Done.

---

_[src/android/org/uproxy/tun2socks/Tunnel.java, line 127 [r2]](https://reviewable.io:443/reviews/uproxy/cordova-plugin-tun2socks/3#-KNsIMSGz6-4BmLAUjqh:-KNsSmfeSO4mP7dCDeOl:-499146009) ([raw file](https://github.com/uproxy/cordova-plugin-tun2socks/blob/55fa490af8903985f4212a350693068441273c22/src/android/org/uproxy/tun2socks/Tunnel.java#L127)):_
<details><summary>

<i>Previously, bemasc (Benjamin M. Schwartz) wrote…</i></summary>



> Why lollipop?  Deserves a comment.
> </details>
> Done.

---

_[src/android/org/uproxy/tun2socks/TunnelManager.java, line 59 [r2]](https://reviewable.io:443/reviews/uproxy/cordova-plugin-tun2socks/3#-KNsIkwxJn4AZRKaY-Zb:-KNsYCfSqq6LwaTvWuX1:-499146009) ([raw file](https://github.com/uproxy/cordova-plugin-tun2socks/blob/55fa490af8903985f4212a350693068441273c22/src/android/org/uproxy/tun2socks/TunnelManager.java#L59)):_
<details><summary>

<i>Previously, bemasc (Benjamin M. Schwartz) wrote…</i></summary>



> Why M?  Needs comment.
> </details>
> Done.

---

_[src/badvpn/tun2socks/tun2socks.c, line 323 [r2]](https://reviewable.io:443/reviews/uproxy/cordova-plugin-tun2socks/3#-KNsJxSdQX-cotYwEzk6:-KNsYd2IkgVEpt8pjuT4:-499146009) ([raw file](https://github.com/uproxy/cordova-plugin-tun2socks/blob/55fa490af8903985f4212a350693068441273c22/src/badvpn/tun2socks/tun2socks.c#L323)):_
<details><summary>

<i>Previously, bemasc (Benjamin M. Schwartz) wrote…</i></summary>



> What's the return value?
> </details>
> Done.

---

_[src/badvpn/tun2socks/tun2socks.c, line 337 [r2]](https://reviewable.io:443/reviews/uproxy/cordova-plugin-tun2socks/3#-KNsJYfP5dao7dxYbGC9:-KNsYheleoH_MzhZ_FH-:-499146009) ([raw file](https://github.com/uproxy/cordova-plugin-tun2socks/blob/55fa490af8903985f4212a350693068441273c22/src/badvpn/tun2socks/tun2socks.c#L337)):_
<details><summary>

<i>Previously, bemasc (Benjamin M. Schwartz) wrote…</i></summary>



> INADDR_ANY is not the same as "any local address".  It's more like "the ANY address".
> </details>
> Done.

---

_[src/badvpn/tun2socks/tun2socks.h, line 48 [r2]](https://reviewable.io:443/reviews/uproxy/cordova-plugin-tun2socks/3#-KNsKXRLQDihiFcBfeGB:-KNs_ybL3Ozac6uWTfJ6:2092268120) ([raw file](https://github.com/uproxy/cordova-plugin-tun2socks/blob/55fa490af8903985f4212a350693068441273c22/src/badvpn/tun2socks/tun2socks.h#L48)):_
<details><summary>

<i>Previously, bemasc (Benjamin M. Schwartz) wrote…</i></summary>



> Is this ever applied to UDP sending?
> </details>
> See my previous comment on DNS/UDP limits.

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/uproxy/cordova-plugin-tun2socks/3)_

**Comment by bemasc** [comment]
Review status: 0 of 19 files reviewed at latest revision, 7 unresolved discussions.

---

_[src/android/org/uproxy/tun2socks/DnsResolverService.java, line 141 [r2]](https://reviewable.io:443/reviews/uproxy/cordova-plugin-tun2socks/3#-KNsDDztCarHvfdckaBq:-KO5agt9UQOTDjlChjt4:775903126) ([raw file](https://github.com/uproxy/cordova-plugin-tun2socks/blob/55fa490af8903985f4212a350693068441273c22/src/android/org/uproxy/tun2socks/DnsResolverService.java#L141)):_
<details><summary>

<i>Previously, albertolalama (alberto lalama) wrote…</i></summary>



> As discussed offline, the DNS over UDP limit is 512 bytes and it's very unlikely that a query will exceed this size. However, the response could have a size of 32767 (max UDP packet size). 
> I'll increase the buffer size to account for this.
> </details>
> Seems like the max UDP packet size is 65535.  (However, I agree that 32K is probably plenty.)

---

_[src/android/org/uproxy/tun2socks/TunnelManager.java, line 59 [r2]](https://reviewable.io:443/reviews/uproxy/cordova-plugin-tun2socks/3#-KNsIkwxJn4AZRKaY-Zb:-KO5dSVpy_WfGPVb5XHw:-1021335763) ([raw file](https://github.com/uproxy/cordova-plugin-tun2socks/blob/55fa490af8903985f4212a350693068441273c22/src/android/org/uproxy/tun2socks/TunnelManager.java#L59)):_
<details><summary>

<i>Previously, albertolalama (alberto lalama) wrote…</i></summary>



> Done.
> </details>
> Hmm.  I don't see a comment.

---

_[src/badvpn/tun2socks/tun2socks.c, line 323 [r2]](https://reviewable.io:443/reviews/uproxy/cordova-plugin-tun2socks/3#-KNsJxSdQX-cotYwEzk6:-KO5eEa1YK3Amx4_9hn_:-942819968) ([raw file](https://github.com/uproxy/cordova-plugin-tun2socks/blob/55fa490af8903985f4212a350693068441273c22/src/badvpn/tun2socks/tun2socks.c#L323)):_
<details><summary>

<i>Previously, albertolalama (alberto lalama) wrote…</i></summary>



> Done.
> </details>
> It looks like it returns the socket's file descriptor number, not the IP address.

---

_[src/badvpn/tun2socks/tun2socks.c, line 337 [r2]](https://reviewable.io:443/reviews/uproxy/cordova-plugin-tun2socks/3#-KNsJYfP5dao7dxYbGC9:-KO5e1YdfjdTw3TAxLau:1467692190) ([raw file](https://github.com/uproxy/cordova-plugin-tun2socks/blob/55fa490af8903985f4212a350693068441273c22/src/badvpn/tun2socks/tun2socks.c#L337)):_
<details><summary>

<i>Previously, albertolalama (alberto lalama) wrote…</i></summary>



> Done.
> </details>
> OK ... this means that you can receive packets from anyone on the internet who manages to send a packet to this port, because ANY includes the public interface(s).  Is that a problem?

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/uproxy/cordova-plugin-tun2socks/3)_

**Comment by alalamav** [comment]
Review status: 0 of 19 files reviewed at latest revision, 7 unresolved discussions.

---

_[src/android/org/uproxy/tun2socks/Tun2Socks.java, line 97 [r2]](https://reviewable.io:443/reviews/uproxy/cordova-plugin-tun2socks/3#-KNsHOMP4MqsrZFx3Vf6:-KO6-9QLUbeunWK0KLM5:-1808023265) ([raw file](https://github.com/uproxy/cordova-plugin-tun2socks/blob/55fa490af8903985f4212a350693068441273c22/src/android/org/uproxy/tun2socks/Tun2Socks.java#L97)):_
<details><summary>

<i>Previously, bemasc (Benjamin M. Schwartz) wrote…</i></summary>



> Nit: I think it would be worth exposing this as a command.  Then maybe we wouldn't need to check the OS version in the JS code; we could just ask the plugin whether it's going to work.
> </details>
> Done. I'll call this method from uProxy once it has been checked in.

---

_[src/android/org/uproxy/tun2socks/TunnelManager.java, line 59 [r2]](https://reviewable.io:443/reviews/uproxy/cordova-plugin-tun2socks/3#-KNsIkwxJn4AZRKaY-Zb:-KO5mNM05orBrX5KGHFV:-1048464391) ([raw file](https://github.com/uproxy/cordova-plugin-tun2socks/blob/55fa490af8903985f4212a350693068441273c22/src/android/org/uproxy/tun2socks/TunnelManager.java#L59)):_
<details><summary>

<i>Previously, bemasc (Benjamin M. Schwartz) wrote…</i></summary>



> Hmm.  I don't see a comment.
> </details>
> This method used to call bindProcessToNetwork. The TargetApi annotation is no longer necessary. I updated the comment of [pluginInitalize](https://github.com/uProxy/cordova-plugin-tun2socks/blob/alalama-changes/src/android/org/uproxy/tun2socks/Tun2Socks.java#L63), which now calls bindProcessToNetwork.

---

_[src/badvpn/tun2socks/tun2socks.c, line 323 [r2]](https://reviewable.io:443/reviews/uproxy/cordova-plugin-tun2socks/3#-KNsJxSdQX-cotYwEzk6:-KO5ns0zCOdYJ60C8doF:2045944120) ([raw file](https://github.com/uproxy/cordova-plugin-tun2socks/blob/55fa490af8903985f4212a350693068441273c22/src/badvpn/tun2socks/tun2socks.c#L323)):_
<details><summary>

<i>Previously, bemasc (Benjamin M. Schwartz) wrote…</i></summary>



> It looks like it returns the socket's file descriptor number, not the IP address.
> </details>
> yes. fixed now.

---

_[src/badvpn/tun2socks/tun2socks.c, line 337 [r2]](https://reviewable.io:443/reviews/uproxy/cordova-plugin-tun2socks/3#-KNsJYfP5dao7dxYbGC9:-KO5o0pA_ZVD0_TtTrE0:1427900090) ([raw file](https://github.com/uproxy/cordova-plugin-tun2socks/blob/55fa490af8903985f4212a350693068441273c22/src/badvpn/tun2socks/tun2socks.c#L337)):_
<details><summary>

<i>Previously, bemasc (Benjamin M. Schwartz) wrote…</i></summary>



> OK ... this means that you can receive packets from anyone on the internet who manages to send a packet to this port, because ANY includes the public interface(s).  Is that a problem?
> </details>
> Although unlikely, it is not impossible for an attacker to inject fake DNS responses. I'm binding the socket to a the loopback address instead.

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/uproxy/cordova-plugin-tun2socks/3)_

**Comment by trevj** [comment]
Great work. I've been running VPN on my own phone, too, with good results.

I'm mostly looking at this from the perspective of maintainability:
- what are the uProxy specific pieces?
- where did the other stuff come from?
- how will we keep up to date with upstream fixes?

I think we should expand the README. Thanks to chatting with @albertolalama and my own poking around, is this the following a fair summary?:

We re-use and have used as a starting point a bunch of open source from [Psiphon](https://psiphon.ca/uz@Latn/open-source.html), specifically https://github.com/mei3am/ps
- `src/android`:
  - starting point: https://github.com/mei3am/ps/tree/master/Android/app
  - `Android/app/src/main/java/ca/psiphon/PsiphonTunnel.java` -> `src/android/org/uproxy/tun2socks/Tunnel.java`
  - `Android/app/src/main/java/com/psiphon3/psiphonlibrary/TunnelManager.java` -> `src/android/org/uproxy/tun2socks/Tunnel.java`
    *`Android/app/src/main/java/com/psiphon3/psiphonlibrary/TunnelVpnService.java` -> `src/android/org/uproxy/tun2socks/TunnelVpnService.java`
- `src/badvpn`:
  - starting point: https://github.com/mei3am/ps/tree/master/Android/badvpn
  - uProxy-specific changes mostly confined to tun2socks/tun2socks.c and marked with `// ==== UPROXY ====` (like Psiphon-specific changes)
- `src/android/libs/jsocks.jar`:
  - http://jsocks.sourceforge.net/

**Comment by alalamav** [comment]
Thank you for compiling a succinct summary, Trevor. I'll make sure to update the README with the code sources.
You raise an interesting question around code maintenance and how to keep up with upstream fixes. I don't have a straight answer at this point, but I think this is something we should discuss with @bemasc.

File: src/android/org/uproxy/tun2socks/DnsResolverService.java [hunk 1]
```diff
@@ -0,0 +1,285 @@
...
+    int separatorIndex = m_socksServerAddress.indexOf(':');
+    String ip = m_socksServerAddress.substring(0, separatorIndex);
+    int port = Integer.parseInt(m_socksServerAddress.substring(separatorIndex + 1));
+    return new InetSocketAddress(ip, port);
+  }
+
+  private class DnsUdpToSocksResolver extends Thread {
+    private static final String LOG_TAG = "DnsUdpToSocksResolver";
+    private static final String DNS_RESOLVER_IP = "8.8.8.8";
+    private static final int MAX_UDP_DATAGRAM_LEN = 32767; // DNS max over UDP
```
Comment (bemasc) [inline src/android/org/uproxy/tun2socks/DnsResolverService.java]: Nit: Based on our conversation today, it sounds like the theoretical max is 64K, and the realistic max is 8K.  I'm fine with 32K too, but I think we should improve the comment regardless.

File: src/badvpn/tun2socks/tun2socks.c [hunk 2]
```diff
@@ -1251,9 +1416,9 @@ int process_device_udp_packet (uint8_t *data, int data_len)
 
             // if transparent DNS is enabled, any packet arriving at out netif
             // address to port 53 is considered a DNS packet
-            is_dns = (options.udpgw_transparent_dns &&
-                      ipv4_header.destination_address == netif_ipaddr.ipv4 &&
-                      udp_header.dest_port == hton16(53));
+            is_dns = (options.transparent_dns &&
+                      // ipv4_header.destination_address == netif_ipaddr.ipv4 &&
```
Comment (bemasc) [inline src/badvpn/tun2socks/tun2socks.c]: Nit: dead code in a comment?

File: src/badvpn/tun2socks/tun2socks.h [hunk 3]
```diff
@@ -44,3 +44,9 @@
 
 // option to override the destination addresses to give the SOCKS server
 //#define OVERRIDE_DEST_ADDR "10.111.0.2:2000"
+
+// maximum number of bytes for udp send/recv
+ #define UDP_MAX_DATAGRAM_BYTES 32767
```
Comment (bemasc) [inline src/badvpn/tun2socks/tun2socks.h]: Nit: as above, needs better comment.

**Comment by bemasc** [comment]
:+1: with nits.  I know you have more improvements queued up, so let's land this and get to the next one!

File: src/badvpn/tun2socks/tun2socks.h (hunk 3 above)
Comment (alalamav) [inline src/badvpn/tun2socks/tun2socks.h]: Done.

File: src/badvpn/tun2socks/tun2socks.c (hunk 2 above)
Comment (alalamav) [inline src/badvpn/tun2socks/tun2socks.c]: I had commented out this line to check if we were receiving non-DNS UDP packets.
It should be uncommented.

File: src/android/org/uproxy/tun2socks/DnsResolverService.java (hunk 1 above)
Comment (alalamav) [inline src/android/org/uproxy/tun2socks/DnsResolverService.java]: Done. As per our discussion, setting the size to 64K.

**Comment by bemasc** [comment]
_[src/android/libs/jsocks.jar, line  [r2]](https://reviewable.io:443/reviews/uproxy/cordova-plugin-tun2socks/3#-KNsBeYdlgRqDSPxQYKm:-KPOS5nhGoEEIxLcq00-:bt2vr0y) ([raw file](https://github.com/uproxy/cordova-plugin-tun2socks/blob/55fa490af8903985f4212a350693068441273c22/src/android/libs/jsocks.jar#L)):_
<details><summary>

<i>Previously, albertolalama (alberto lalama) wrote…</i></summary>



> The JAR is 395 KB. The source code is in SourceForge, but seems like it's not actively maintained. We could create our fork, and include the JAR ad hoc. But I don't know if this is a better alternative.
> </details>
> Fixed!

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/uproxy/cordova-plugin-tun2socks/3)_

**Comment by bemasc** [comment]
:lgtm: with one tiny nit

---

Review status: 0 of 18 files reviewed at latest revision, 1 unresolved discussion.

---

_[src/android/org/uproxy/tun2socks/DnsResolverService.java, line 91 [r4]](https://reviewable.io:443/reviews/uproxy/cordova-plugin-tun2socks/3#-KNs2scs_PHxH9W7vPEB-r4-91:-KPORmqgIVJX2A30MTx6:bt6tvv1) ([raw file](https://github.com/uproxy/cordova-plugin-tun2socks/blob/86dd5ed9f88a9ad8ca6196553ac9ca5a8dcd431a/src/android/org/uproxy/tun2socks/DnsResolverService.java#L91)):_
<details><summary>

<i>Previously, albertolalama (alberto lalama) wrote…</i></summary>



> Done. As per our discussion, setting the size to 64K.
> </details>
> Nit: This is not related to Google DNS.  The DNS TCP protocol only has two bytes for length, so larger packets are cannot even be represented.

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/uproxy/cordova-plugin-tun2socks/3#-:-KPOSRapHk3q7_3tav5A:b-qytssi)_

**Comment by alalamav** [comment]
Review status: 0 of 18 files reviewed at latest revision, 2 unresolved discussions.

---

_[src/android/org/uproxy/tun2socks/DnsResolverService.java, line 91 [r4]](https://reviewable.io:443/reviews/uproxy/cordova-plugin-tun2socks/3#-KNs2scs_PHxH9W7vPEB-r4-91:-KPOUsFnnXrh0Btuh2m_:b-896fix) ([raw file](https://github.com/uproxy/cordova-plugin-tun2socks/blob/86dd5ed9f88a9ad8ca6196553ac9ca5a8dcd431a/src/android/org/uproxy/tun2socks/DnsResolverService.java#L91)):_
<details><summary>

<i>Previously, bemasc (Benjamin M. Schwartz) wrote…</i></summary>



> Nit: This is not related to Google DNS.  The DNS TCP protocol only has two bytes for length, so larger packets are cannot even be represented.
> </details>
> Done.

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/uproxy/cordova-plugin-tun2socks/3)_

**Pull request merged**